doc/xfpt.1
.TH XFPT 1
.SH NAME
xfpt - make XML from plain text
.SH SYNOPSIS
.B xfpt [options] [source file]
.
.SH DESCRIPTION
.rs
.sp
\fBxfpt\fP converts a marked up text file to XML. The markup is simple and
consists of lines that begin with a dot ("directive lines") and sequences in
the text that begin with an ampersand ("flags"). The flag sequences are not
built-in, but are defined by directives. As well as the in-built directives,
macros can be used to implement higher level concepts. A standard set of macros 
and flags that generate DocBook XML is provided. There is a full description in
the \fBxfpt\fP specification, which is distributed as a PDF file, an HTML file, 
and as \fBxfpt\fP source.
.
.SH OPTIONS
.rs
.TP 10
\fB-help\fP or \fB--help\fP
This causes \fBxfpt\fP to output its "usage" information and then exit.
.TP
\fB-o\fP \fIfile\fP
This specifies the output file. The default is the standard output if no source 
file is given, and the source file name with a \fI.xml\fP extension otherwise.
.TP
\fB-S\fP \fIdirectory\fP
This specifies an alternative "share" directory in which to find standard 
configuration files.
.TP
\fB-v\fP or \fB--version\fP
Output the \fBxfpt\fP version and exit.
.
.SH AUTHOR
.rs
.sp
Philip Hazel
.br
Retired from University Computing Service
.br
Cambridge, England.
.P
.in 0
Last updated: 08 March 2023
.br
Copyright (c) 2023 University of Cambridge.


ChangeLog
ChangeLog file for xfpt
-----------------------

Release 1.00 08-March-2023
--------------------------

1. Fix crash when --version used (double file close, a consequence of 4 below).

2. Some general code tidies have been done and the tests have been extended in
   the light of coverage checking and to test various command line options. The
   documentation has also been given an overhaul. It seems like time to move to
   a 1.00 release.


Release 0.11 04-September-2020
------------------------------

1. Fix a buffer overrun (read) for a missing delimiter after $=2 (for example).

2. Give an error if a binary zero is encountered in the input, and ignore it.

3. Give an error if input source nesting is too deep (e.g. a runaway recursive
macro).

4. Yet more tidying to ensure all memory gets freed (2 below was very
incomplete).


Release 0.10 28-August-2018
---------------------------

1. A few minor code re-writings to avoid compiler warnings when compiled with
   lots of warnings turned on.

2. Slightly more involved re-writings to ensure that all memory gets freed.

3. Fixed typos in the documentation.

4. Recognize --version as well as -v.


Release 0.09 18-May-2012
------------------------

1. Add the "check" feature to .push to give a warning for certain items on the
   stack that are automatically popped at the end of processing.

2. Update the standard macros so that "check" warnings are given for unclosed
   lists, displays, footnotes, etc.


Release 0.08 22-July-2009
-------------------------

1. The macro ".image" was generating </imagedata> on a new line, thus causing
   some XML processors to think there was content (the newline) when there is
   not supposed to be.

2. A "read random memory" bug that was probably introduced with change 0.06/1
   caused a crash on 64-bit Mac OS X, though not in other environments. The
   random memory was allocated, so never showed with valgrind. The crash
   happened right at the end of the input.


Release 0.07
------------

Added macros .endpreface, etc, and .pi. There is no change of code, except to
bump the version number.


Release 0.06
------------

1. The .include directive was not working correctly when it appeared inside
   a macro. The contents of the file were not read until after the macro's
   lines had been processed. Mixing of .include and macro calls, in any
   combination, should now work correctly. A full traceback of the position of
   any errors is now given (formerly it stopped when a position in a file was
   given).


Release 0.05
------------

1. The .nonl directive was not being handled correctly when it appeared at the
   start of a paragraph and in some other cases. Its argument is now handled as
   the next input line, but without a newline on the end.


Release 0.04
------------

1. Single quotes were being turned into typographic quotes inside <literal>
   phrases, when they should not be. Normal cases (where <literal> is generated
   by an xfpt macro) are now handled correctly.

2. The .vlist macro didn't set .literal off at the end, causing a following
   .new macro not to work properly; it generated bad XML.


Release 0.03
------------

1. Added .nest built-in directive in order to support .footnote and .endnote
   macros (also added).

2. Other new macros: .figure, .image, .preface, .appendix, .colophon, .table.


Release 0.02
------------

1. Added the .inliteral directive.

2. Added the .nonl directive.

3. Updated the .new and .wen directives to work properly in literal sections.


Release 0.01
------------

Added new macros .index-from, .index-to, .index-see, and .index-seealso.


Release 0.00
------------

This is the first release. There are no changes yet!

****


src/xfpt.c
/*************************************************
*     xfpt - Simple ASCII->Docbook processor     *
*************************************************/

/* Copyright (c) University of Cambridge, 2023 */
/* Written by Philip Hazel, started in 2006 */

/* This module contains the main program and initialization functions. */

#include "xfpt.h"



/*************************************************
*                 Static variables               *
*************************************************/

static uschar *xfpt_filename = NULL;
static uschar *out_filename = NULL;
static BOOL out_filename_onheap = FALSE;



/*************************************************
*                  Usage                         *
*************************************************/

static void
usage(void)
{
(void)fprintf(stderr,
  "Usage: xfpt [[-]-help]\n"
  "            [-o <output-file>]\n"
  "            [-S <share-directory>]\n"
  "            [-v or --version]\n"
  "            [input-file]\n");
}



/*************************************************
*          Command line argument decoding        *
*************************************************/

/* Arguments: as for main()
   Returns:   TRUE if OK
*/

static BOOL
xfpt_decode_arg(int argc, char **argv)
{
int i;
for (i = 1; i < argc; i++)
  {
  uschar *arg = US argv[i];
  if (*arg != '-') break;
  if (Ustrcmp(arg, "-o") == 0)
    {
    out_filename = US argv[++i];
    if (out_filename == NULL) { usage(); return FALSE; }
    }
  else if (Ustrcmp(arg, "-S") == 0)
    {
    xfpt_share = US argv[++i];
    if (xfpt_share == NULL) { usage(); return FALSE; }
    }
  else if (Ustrcmp(arg, "-help") == 0 || Ustrcmp(arg, "--help") == 0)
    {
    usage();
    return FALSE;
    }
  else if (Ustrcmp(arg, "-v") == 0 || Ustrcmp(arg, "--version") == 0)
    {
    (void)fprintf(stdout, "xpft version %s\n", xfpt_version);
    exit(0);
    }
  else
    {
    (void)fprintf(stderr, "xfpt: unknown option \"%s\"\n", arg);
    usage();
    return FALSE;
    }
  }

/* Require there to be either 0 or 1 command line argument left. */

if (argc > i + 1)
  {
  usage();
  return FALSE;
  }

/* This will set NULL if there is no file name. If there is a file name and no
output file is specified, default it to the input name with a .xml extension. */

xfpt_filename = US argv[i];
if (xfpt_filename != NULL && out_filename == NULL)
  {
  uschar *p;
  int len = Ustrlen(xfpt_filename);
  out_filename = misc_malloc(len + 5);
  out_filename_onheap = TRUE;
  Ustrcpy(out_filename, xfpt_filename);
  if ((p = Ustrrchr(out_filename, '.')) != NULL) len = p - out_filename;
  Ustrcpy(out_filename + len, ".xml");
  }

return TRUE;
}



/*************************************************
*          Free a tree node and its children     *
*************************************************/

static void
free_tree(tree_node *t)
{
if (t == NULL) return;
free_tree(t->left);
free_tree(t->right);
free(t->data);
free(t);
}



/*************************************************
*          Clean up function at exit             *
*************************************************/

static void
clean_up(void)
{
int warnpop = 0;
macrodef *md = macrolist;
macroexe *me = macrocurrent;
istackstr *ist = istack;
flagstr *f = flaglist;

/* Empty the stack of pushed texts. This must be done first because it calls
para_process(), which needs access to flags, entities, etc. */

while (pushed != NULL)
  {
  pushstr *temp;
  if (!suppress_warnings)
    {
    if (pushed->check != 0)
      {
      if (warnpop++ == 0)
        fprintf(stderr,
           "** Warning: one or more items were left unclosed at the end of processing.\n"
           "   The numbers are the lines in the original file %s from where\n"
           "   the items were generated:\n",
           ((xfpt_filename == NULL)? "(stdin)" : (char *)xfpt_filename));
      if (pushed->macname == NULL)
        fprintf(stderr, "%d: %s\n", pushed->check, pushed->string);
      else
        fprintf(stderr, "%d: .%s\n", pushed->check, pushed->macname);
      if (warnpop > 10)
        {
        fprintf(stderr, "... too many to list\n");
        suppress_warnings = TRUE;
        }
      }
    }
  para_process(pushed->string);
  (void)fprintf(outfile, "\n");
  temp = pushed;
  pushed = pushed->next;
  free(temp);
  }

/* Release macro definitions */

while (md != NULL)
  {
  macrodef *mdnext = md->next;
  argstr *as = md->args;

  while (as != NULL)
    {
    argstr *next = as->next;
    free(as->string);
    free(as);
    as = next;
    }

  as = md->lines;
  while (as != NULL)
    {
    argstr *next = as->next;
    free(as->string);
    free(as);
    as = next;
    }

  free(md->name);
  free(md);
  md = mdnext;
  }

/* Release macro execution blocks */

while (me != NULL)
  {
  macroexe *menext = me->prev;
  macro_free(me);
  me = menext;
  }

/* Release input source blocks */

while (ist != NULL)
  {
  istackstr *prev = ist->prev;
  if (ist->file != NULL && ist->file != stdin) fclose(ist->file);
  free(ist);
  ist = prev;
  }

/* Release flag definitions */

while (f != NULL)
  {
  flagstr *fnext = f->next;
  flag_free(f);
  f = fnext;
  }

/* Free entity definitions */

free_tree(entities);

/* Release general buffers, close output unless stdout, and free its name if
necessary. */

free(inbuffer);
free(parabuffer);
if (outfile != NULL && outfile != stdout) fclose(outfile);
if (out_filename_onheap) free(out_filename);
}



/*************************************************
*          Entry point and main program          *
*************************************************/

int
main(int argc, char **argv)
{
BOOL para_unfinished[MAXNEST+1];
uschar *p, *q;

atexit(clean_up);

if (!xfpt_decode_arg(argc, argv)) return EXIT_FAILURE;

inbuffer = misc_malloc(INBUFFSIZE);
parabuffer = misc_malloc(PARABUFFSIZE);

/* Set up the first file */

istackbase = istack = misc_malloc(sizeof(istackstr));
istack->prev = NULL;
istack->linenumber = 0;

from_type_ptr = 0;
from_type[from_type_ptr] = FROM_FILE;

if (xfpt_filename == NULL)
  {
  istack->file = stdin;
  Ustrcpy(istack->filename, US"(stdin)");
  }
else
  {
  Ustrcpy(istack->filename, xfpt_filename);
  istack->file = Ufopen(xfpt_filename, "rb");
  if (istack->file == NULL)
    error(0, istack->filename, strerror(errno));    /* Hard */
  }

/* Set up the output file. */

if (out_filename == NULL || Ustrcmp(out_filename, "-") == 0)
  {
  outfile = stdout;
  }
else
  {
  outfile = Ufopen(out_filename, "wb");
  if (outfile == NULL)
    error(0, out_filename, strerror(errno));   /* Hard error */
  }

/* Process the input */

nest_level = 0;
para_unfinished[0] = FALSE;

while ((p = read_nextline()) != NULL)
  {
  if (*p == '.') dot_process(p); else switch (literal_state)
    {
    case LITERAL_LAYOUT:
    para_process(p);
    break;

    case LITERAL_TEXT:  /* Escape & < and > and copy anything else */
    while (*p != 0)
      {
      int c = *p++;
      if (c == '&')      (void)fprintf(outfile, "&amp;");
      else if (c == '<') (void)fprintf(outfile, "&lt;");
      else if (c == '>') (void)fprintf(outfile, "&gt;");
      else (void)fputc(c, outfile);
      }
    break;

    case LITERAL_XML:
    (void)fprintf(outfile, "%s", CS p);
    break;

    default:
    case LITERAL_OFF:
    q = p;
    while (isspace(*q)) q++;
    if (*q != 0)
      {
      int nest_info;
      p = read_paragraph(p, &nest_info);
      if (!para_unfinished[nest_level])
        {
        (void)fprintf(outfile, "<");
        para_process(US"para&xfpt.rev;");
        (void)fprintf(outfile, ">\n");
        }

      para_process(p);
      if (nest_info == NEST_BEGIN)
        {
        if (nest_level >= MAXNEST) error(27); else
          {
          nest_literal_stack[nest_level] = literal_state;
          para_unfinished[nest_level++] = TRUE;
          }
        }
      else (void)fprintf(outfile, "</para>\n");

      para_unfinished[nest_level] = FALSE;

      if (nest_info == NEST_END)
        {
        if (nest_level <= 0) error(28);
          else literal_state = nest_literal_stack[--nest_level];
        }
      }
    break;
    }
  }

/* Clean up of heap memory and file closure happens in the clean_up() function
set up by atexit(). */

return return_code;
}

/* End of xfpt.c */
